Issue #14 - Spring Rest Exception Handler

gala-lightfoot opened the pull request:
We now throw a generic EntityNotFoundException when a requested mission is not found, as opposed to returning a 404 response entity.  This is caught by the RestExceptionHandler, which returns a custom response and 404 error code.

RestExceptionHandler intercepts exceptions thrown from within **all** Spring controllers (if a handler is defined for that particular exception).

Happy to discuss the Spring alternatives for exception handling if we want this to go in another direction.

gala-lightfoot committed:
Nonexistant mission controller call now throws REST exception

gala-lightfoot committed:
Utilise a Spring Exception Handler

gala-lightfoot committed:
Comments

gala-lightfoot committed:
Comments

gala-lightfoot committed:
Imports

gala-lightfoot committed:
Comments

gala-dan commented on src/main/java/org/galatea/starter/ApiError.java:
```diff
@@ -0,0 +1,24 @@
+import lombok.ToString;
+import org.springframework.http.HttpStatus;
+
+@Getter
```
Don't need all of these annotations. [@Data](https://projectlombok.org/features/Data) is a convenience annotation which provides some of the others. Looking at how ApiError is used I suggest you need @Getter, @AllArgsConstructor and maybe @ToString.

gala-dan commented on src/test/java/org/galatea/starter/entrypoint/SettlementRestControllerTest.java:
```diff
@@ -138,12 +141,16 @@ public void testGetMissionFound() throws Exception {
   public void testGetMissionNotFound() throws Exception {
     long msnId = 1091L;
 
+    SettlementMissionNotFoundException exc = new SettlementMissionNotFoundException(msnId);
```
Use the name exception rather than exc here.

gala-lightfoot committed:
ApiError annotations.  Test exception naming.

gala-lightfoot committed:
Merge branch 'develop' into feature/issue-14-spring-rest-exceptions

paulo-reichert commented on src/main/java/org/galatea/starter/entrypoint/SettlementRestController.java:
```diff
@@ -85,7 +85,7 @@
     }
 
-    return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+    throw new SettlementMissionNotFoundException(id);
```
I don't agree that if we don't find an object in the database we should throw an exception. Exceptions should represent issues in the execution of the application that were unexpected and that workflow has to stop.

paulo-reichert commented on src/main/java/org/galatea/starter/domain/exception/SettlementMissionNotFoundException.java:
```diff
@@ -0,0 +1,9 @@
+package org.galatea.starter.domain.exception;
+
+public class SettlementMissionNotFoundException extends RuntimeException {
```
This exception seems to be oddly specific. If we consider an application that may have tens of domain objects and various error conditions in its logic, we would end up far too many exceptions. We should look for the common errors that may arise in the code and model exceptions around those.

gala-lightfoot committed:
Handle DataAccessException.  Define generic EntityNotFoundException.

gala-lightfoot committed:
Merge branch 'develop' of github.com:GalateaRaj/fuse-starter-java into feature/issue-14-spring-rest-exceptions

gala-lightfoot committed:
Merge branch 'feature/issue-14-spring-rest-exceptions' of github.com:GalateaRaj/fuse-starter-java into feature/issue-14-spring-rest-exceptions

gala-lightfoot committed:
No longer return ResponseEntity from SettlementRestController.
Return client-friendly error messages for EntityNotFoundException, DataAccessException & HttpMessageNotReadableException

gala-lightfoot committed:
Inline missionIds return

paulo-reichert commented on src/main/java/org/galatea/starter/ApiError.java:
```diff
@@ -0,0 +1,21 @@
+@Getter
+@ToString
+@AllArgsConstructor
+public class ApiError {
```
Do we need to have this class? It seems to be used only in the exception handler to communicate the HTTP response code to be returned to the private `buildResponseEntity()` method so in that case what is the downside of just having the `HttpStatus` and message passed in without the wrapper?

paulo-reichert replied:
Should this move to the service package? The root package is mostly for the application starting point and configs.

gala-lightfoot committed:
Comments

paulo-reichert committed:
Merge branch 'develop' into feature/issue-14-spring-rest-exceptions

gala-lightfoot committed:
Move ApiError to service package

paulo-reichert commented on src/main/java/org/galatea/starter/domain/exception/EntityNotFoundException.java:
```diff
@@ -0,0 +1,9 @@
+package org.galatea.starter.domain.exception;
+
+public class EntityNotFoundException extends RuntimeException {
```
This exception is likely to be used only in the RestController so I think it should not be in the domain package, but closer to where it will actually be used.

paulo-reichert commented on src/main/java/org/galatea/starter/service/RestExceptionHandler.java:
```diff
@@ -0,0 +1,51 @@
+ */
+@ControllerAdvice
+@Slf4j
+public class RestExceptionHandler {
```
This should be in the entrypoint package, given that it relates exclusively to rest services.

paulo-reichert commented on src/main/java/org/galatea/starter/service/ApiError.java:
```diff
@@ -0,0 +1,21 @@
+@Getter
+@ToString
+@AllArgsConstructor
+public class ApiError {
```
This should be actually moved to the entrypoint package given that, like the RestExceptionHandler, it applies only to rest services.

gala-lightfoot committed:
File restructuring

paulo-reichert committed:
Merge branch 'develop' into feature/issue-14-spring-rest-exceptions

paulo-reichert reviewed (approved):
Good to go.
